protocols/request-response/tests/error_reporting.rs

```diff
@@ -0,0 +1,727 @@
+use async_std::channel;
+use async_std::future::timeout;
+use async_std::task::{sleep, spawn};
+use async_trait::async_trait;
+use futures::prelude::*;
+use libp2p_request_response as request_response;
+use libp2p_request_response::ProtocolSupport;
+use libp2p_swarm::{StreamProtocol, Swarm};
+use libp2p_swarm_test::SwarmExt;
+use request_response::{Codec, InboundFailure, OutboundFailure};
+use std::time::Duration;
+use std::{io, iter};
+
+#[async_std::test]
+async fn report_outbound_failure_on_read_response() {
+    let _ = env_logger::try_init();
+
+    let protocols = iter::once((StreamProtocol::new("/test/1"), ProtocolSupport::Full));
+    let cfg = request_response::Config::default();
+
+    let mut swarm1 = Swarm::new_ephemeral(|_| {
+        request_response::Behaviour::<TestCodec>::new(protocols.clone(), cfg.clone())
+    });
+    let peer1_id = *swarm1.local_peer_id();
+
+    let mut swarm2 =
+        Swarm::new_ephemeral(|_| request_response::Behaviour::<TestCodec>::new(protocols, cfg));
+    let peer2_id = *swarm2.local_peer_id();
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
```

Comment:
Can we extract all of that into a helper fn to remove some duplication?

Reply:
If you agree on [this](https://github.com/libp2p/rust-libp2p/blob/21a34fc719531d0536f5c95fa628c9d6c6544f3d/protocols/request-response/tests/error_reporting.rs#L18-L66), I will change other tests too.

Reply:
[with join instead of spawn](https://github.com/libp2p/rust-libp2p/blob/0f27d04a4b04788fe940b027de3ca4c301650d48/protocols/request-response/tests/error_reporting.rs#L18-L67)

Reply:
Rewrote all the test. They are much cleaner now.